src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs

```diff
@@ -545,22 +537,45 @@ private void Poll()
             }
         }
 
-        private List<ConsumeResult<K, V>> PollKafka(CancellationToken token)
+        private void ProcessExceptions(Exception exception)
+        {
+            if (exception == null)
+                return;
+
+            var directive = _decider(exception);
+            ProcessError(exception);
+            if (directive == Directive.Resume)
+                return;
+            
+            _pollCancellation?.Cancel();
+            if(directive == Directive.Stop && _log.IsErrorEnabled)
+                _log.Error(exception, "Exception when polling from consumer, stopping actor: {0}", exception.Message);
+            Context.Stop(Self);
+        }
+
+        private (List<ConsumeResult<K, V>>, Exception) PollKafka(CancellationToken token)
         {
             ConsumeResult<K, V> consumed = null;
             var i = 10; // 10 poll attempts
             var timeout = Math.Max((int) _pollTimeout.TotalMilliseconds / i, 1);
-            var pooled = new List<ConsumeResult<K, V>>();
+            var polled = new List<ConsumeResult<K, V>>();
             do
```

Comment:
New polling method. Instead of discarding all the consumed message when an exception is thrown, we return both the polled messages before the exception and the exception. We then first push all of the consumed messages and then process the exception. This makes polling consistent on both slow and fast brokers and allows us to implement correct stage restart behavior.